drakrun/drakrun/test/test_vm.py

```diff
@@ -0,0 +1,269 @@
+        # destroy_vm(test_vm.vm_name)
+        test_vm.create(pause=True)
+        assert get_vm_state(test_vm.vm_name) == 'p'
+
+        logging.info("test vm save with pause=True")
+        test_vm.save(snapshot_file, pause=True)
+        assert get_vm_state(test_vm.vm_name) == 'p'
+
+        # should destroy the vm
+        logging.info("test vm save with with no pause/cont args")
+        test_vm.save(snapshot_file)
+        with pytest.raises(Exception):
+            get_vm_state(test_vm.name)
+
+    def test_vm_pause(self, test_vm):
+        # initialize the VM after previous destruction
+        test_vm.create()
+
+        assert get_vm_state(test_vm.vm_name) != 'p'
+
+        # test-hvm64-example goes to shutdown immediately, we get `--ps--` state during assertion
+
+        # logging.info("testing pause on VM")
+        # test_vm.pause()
+        # assert get_vm_state(test_vm.vm_name) == 'p'
+
+        # manual test shows, xl pause on a paused VM doesn't give any errors but pauses the VM again
+        # requiring the VM be unpaused twice for reaching running state
+
+        # logging.info("testing pause on a paused vm VM")
+        # with pytest.raises(Exception):
+        #     test_vm.pause()
+
+        destroy_vm(test_vm.vm_name)
+
+    def test_vm_restore(self, config, snapshot_file, test_vm):
+        # if snapshot doesn't exist
+        logging.info("test vm restore without snapshot file")
+        with remove_files([snapshot_file]):
+            with pytest.raises(Exception):
+                test_vm.restore(snapshot_path=snapshot_file)
+                assert test_vm.is_running is False
+
+        # if configuration file doesn't exist
+        logging.info("test vm restore without config")
+        with remove_files([config]):
+            with pytest.raises(Exception):
+                test_vm.restore(snapshot_path=snapshot_file)
+                assert test_vm.is_running is False
+
+        # although test-hvm64-example doesn't depend on storage backend
+        # some test like this would have been good where storage backend doesn't exist
+        # and it is trying to restore from vm-1 or vm-0
+        # vm-0 should fail but vm-1 should succeed
+        # if backend.exists_vm(0) is False:
+        #     with pytest.raises(Exception):
+        #         test_vm.restore()
+        #         assert test_vm.is_running is False
+
+        # should not raise any exceptions if everything is fine
+        logging.info("test vm with proper args")
+        test_vm.restore(snapshot_path=snapshot_file)
+        assert get_vm_state(test_vm.vm_name) != 'p'
+
+        destroy_vm(test_vm.vm_name)
+
+        logging.info("test vm with proper args and pause=True")
+        test_vm.restore(snapshot_path=snapshot_file, pause=True)
+        assert get_vm_state(test_vm.vm_name) == 'p'
+
+        # restoring a restored VM
+        # what should be the expected behavior?
+        # test_vm.restore(cfg_path= config, snapshot_path = snapshot_file)
```

Comment:
As I've mentioned in some other comment it's convenient to create a new, fresh VM without throwing.